### PR TITLE
 executor/split: return split result when do split region and refine split timeout logic.

### PR DESCRIPTION
**Pull request by crazycs520**
### What problem does this PR solve? 
* return a split result when do split region
* refine split timeout logic.

The old split region timeout logic was return error when timeout. But return error is not suitable. Because after TiDB send `split` and `scatter` request to PD, we should think the request must be executed successfully. The cost time of `split` and `scatter` request finish depends on the TiKV/PD cluster whether busy.

```SQL
mysql>set @@tidb_wait_split_region_timeout=10;
Query OK, 0 rows affected
Time: 0.000s
mysql>split table t between (100000) and (200000) regions 10;
+--------------------+----------------------+
| TOTAL_SPLIT_REGION | SCATTER_FINISH_RATIO |
+--------------------+----------------------+
| 9                  | 0.444444444444       |
+--------------------+----------------------+
1 row in set
Time: 10.427s
```

### Check List 

Tests 

 - Unit test
 - Manual test (add detailed scripts or steps below)

Code changes

 - Has exported function/method change

Side effects

Related changes

 - Need to cherry-pick to the release branch

### CODE REVIEW

File: executor/executor_test.go [hunk 1]
```diff
@@ -4077,7 +4075,7 @@ func (s *testSuite) TestShowTableRegion(c *C) {
 
 	// Test show table regions.
 	tk.MustExec(`set @@session.tidb_wait_split_region_finish=1;`)
-	tk.MustExec(`split table t_regions between (0) and (10000) regions 4;`)
+	tk.MustQuery(`split table t_regions by (2500),(5000),(7500);`).Check(testkit.Rows("3 1"))
```
Comment (bb7133) [inline executor/executor_test.go]: why changed the split stmt from `between ...` to `by (2500),(5000),(7500)` here? Is it possible to use this query with `between`?

File: executor/split.go [hunk 2]
```diff
@@ -74,28 +103,44 @@ func (e *SplitIndexRegionExec) Next(ctx context.Context, _ *chunk.Chunk) error {
...
+	e.splitRegions = len(regionIDs)
 	if !e.ctx.GetSessionVars().WaitSplitRegionFinish {
 		return nil
 	}
+	remainMillisecond := 0
+	finishScatterNum := 0
 	for _, regionID := range regionIDs {
-		err := s.WaitScatterRegionFinish(regionID)
-		if err != nil {
+		if isCtxDone(ctxWithTimeout) {
```
Comment (bb7133) [inline executor/split.go]: The code is confusing, please add some comments here.

File: executor/split.go [hunk 3]
```diff
@@ -244,50 +309,73 @@ func (e *SplitTableRegionExec) Next(ctx context.Context, _ *chunk.Chunk) error {
...
 			}
 		})
-
 		if isCtxDone(ctxWithTimeout) {
-			return errors.Errorf("wait split region scatter timeout(%v)", e.ctx.GetSessionVars().GetSplitRegionTimeout())
+			// Do not break here for checking remain region scatter finished with a very short back off time.
+			// Imagine this situation, we split region 1,2,3, and timeout on wait region 1 scatter,
+			// but region 2 and region 3 is already scatter finish, then we should return result finish scatter region num 2,
+			// instead of finish scatter region num 0.
+			remainMillisecond = 50
```
Comment (bb7133) [inline executor/split.go]: It's better to define a constant instead of `50`

File: executor/split.go [hunk 4]
```diff
@@ -74,28 +103,44 @@ func (e *SplitIndexRegionExec) Next(ctx context.Context, _ *chunk.Chunk) error {
...
+	finishScatterNum := 0
 	for _, regionID := range regionIDs {
-		err := s.WaitScatterRegionFinish(regionID)
-		if err != nil {
+		if isCtxDone(ctxWithTimeout) {
+			break
+		}
+
+		if isCtxDone(ctxWithTimeout) {
+			// Do not break here for checking remain region scatter finished with a very short back off time.
```
Comment (bb7133) [inline executor/split.go]: Please fix the grammar in this comment.

File: executor/executor_test.go (hunk 1 above)
Comment (crazycs520) [reply executor/executor_test.go]: I just want to test more case. The `between ... and ...` syntax already test in line#4028.

File: executor/split.go (hunk 2 above)
Comment (crazycs520) [reply executor/split.go]: Oh, sorry, It is my mistake.

File: executor/split.go (hunk 4 above)
Comment (crazycs520) [reply executor/split.go]: done.

File: executor/split.go (hunk 3 above)
Comment (crazycs520) [reply executor/split.go]: done.

File: executor/split.go [hunk 5]
```diff
@@ -48,10 +48,37 @@ type SplitIndexRegionExec struct {
...
+	chk.Reset()
+	if e.done {
+		return nil
+	}
+	appendSplitRegionResultToChunk(chk, e.splitRegions, e.finishScatterNum)
+	e.done = true
+	return nil
+}
+
+// checkScatterRegionFinishBackOff is the back off that uses to check region scatter finished when split region timeout.
```
Comment (bb7133) [inline executor/split.go]:
```suggestion
// checkScatterRegionFinishBackOff is the back off time that used to check if a region has finished scattering before split region timeout.
```

File: executor/split.go [hunk 6]
```diff
@@ -48,10 +48,37 @@ type SplitIndexRegionExec struct {
...
+const checkScatterRegionFinishBackOff = 50
+
+// splitIndexRegion uses to split index regions.
```
Comment (bb7133) [inline executor/split.go]:
```suggestion
// splitIndexRegion is used to split index regions.
```

File: executor/split.go [hunk 7]
```diff
@@ -225,16 +269,36 @@ type SplitTableRegionExec struct {
...
+		return nil
+	}
+	appendSplitRegionResultToChunk(chk, e.splitRegions, e.finishScatterNum)
+	e.done = true
+	return nil
 }
 
 // Next implements the Executor Next interface.
-func (e *SplitTableRegionExec) Next(ctx context.Context, _ *chunk.Chunk) error {
+func (e *SplitTableRegionExec) splitTableRegion(ctx context.Context) error {
```
Comment (bb7133) [inline executor/split.go]: Please fix the comment

File: planner/core/planbuilder.go [hunk 8]
```diff
@@ -1002,6 +1002,13 @@ func buildTableRegionsSchema() *expression.Schema {
 	return schema
 }
 
+func buildSplitRegionsSchema() *expression.Schema {
+	schema := expression.NewSchema(make([]*expression.Column, 0, 10)...)
```
Comment (bb7133) [inline planner/core/planbuilder.go]: why making the slice with `10` capacity?

File: executor/split.go [hunk 9]
```diff
@@ -74,28 +106,40 @@ func (e *SplitIndexRegionExec) Next(ctx context.Context, _ *chunk.Chunk) error {
...
 		return nil
 	}
+	remainMillisecond := 0
+	finishScatterNum := 0
 	for _, regionID := range regionIDs {
-		err := s.WaitScatterRegionFinish(regionID)
-		if err != nil {
+		if isCtxDone(ctxWithTimeout) {
+			// Do not break here for checking remain region scatter finished with a very short back off time.
+			// Imagine this situation, we split region 1,2,3, and timeout on wait region 1 scatter,
```
Comment (bb7133) [inline executor/split.go]: IMHO the comment is still not good enough... would you please get help from some other colleagues, like i18n team?

File: executor/split.go [hunk 10]
```diff
@@ -244,50 +308,68 @@ func (e *SplitTableRegionExec) Next(ctx context.Context, _ *chunk.Chunk) error {
...
 	if !e.ctx.GetSessionVars().WaitSplitRegionFinish {
 		return nil
 	}
+
+	remainMillisecond := 0
+	finishScatterNum := 0
 	for _, regionID := range regionIDs {
-		err := s.WaitScatterRegionFinish(regionID)
-		if err != nil {
+		if isCtxDone(ctxWithTimeout) {
```
Comment (winkyao) [inline executor/split.go]: don't repeat yourself.

Comment (crazycs520) [reply executor/split.go]: done.

File: planner/core/planbuilder.go (hunk 8 above)
Comment (crazycs520) [reply planner/core/planbuilder.go]: Sorry, It's a mistake, already fixed.

File: executor/split.go (hunk 7 above)
Comment (crazycs520) [reply executor/split.go]: done.

File: store/tikv/split_region.go [hunk 11]
```diff
@@ -103,10 +103,15 @@ func (s *tikvStore) scatterRegion(regionID uint64) error {
 }
 
 // WaitScatterRegionFinish implements SplitableStore interface.
-func (s *tikvStore) WaitScatterRegionFinish(regionID uint64) error {
+// backOff is the back off time of the wait scatter region.(Milliseconds)
+// if backOff <= 0, will use the default wait scatter back off time.
```
[the remark this answers is not in the document]
Comment (crazycs520) [reply store/tikv/split_region.go]: done.

File: executor/split.go (hunk 9 above)
Comment (crazycs520) [reply executor/split.go]: done.

File: executor/split.go [hunk 12]
```diff
@@ -244,48 +287,79 @@ func (e *SplitTableRegionExec) Next(ctx context.Context, _ *chunk.Chunk) error {
...
-			return errors.Errorf("split region timeout(%v)", e.ctx.GetSessionVars().GetSplitRegionTimeout())
-		}
 	}
+	e.splitRegions = len(regionIDs)
 	if !e.ctx.GetSessionVars().WaitSplitRegionFinish {
 		return nil
 	}
+
+	finishScatterNum := waitScatterRegionFinish(e.ctx, ctxWithTimeout, start, s, regionIDs, e.tableInfo.Name.L, "")
+	e.finishScatterNum = finishScatterNum
```
Comment (bb7133) [inline executor/split.go]:
```
	finishScatterNum := waitScatterRegionFinish(e.ctx, ctxWithTimeout, start, s, regionIDs, e.tableInfo.Name.L, "")
	e.finishScatterNum = finishScatterNum
```
Please combine them in 1 line, the code length simply increase by 2...(from 116 to 118)

File: executor/split.go [hunk 13]
```diff
@@ -225,16 +248,36 @@ type SplitTableRegionExec struct {
...
+	chk.Reset()
+	if e.done {
+		return nil
+	}
+	appendSplitRegionResultToChunk(chk, e.splitRegions, e.finishScatterNum)
+	e.done = true
+	return nil
+}
+
+// splitTableRegion is used to split table regions.
```
Comment (bb7133) [inline executor/split.go]: Mmmm...the comment is just meaningless, and it's not required...Please just remove it.

Comment (crazycs520) [reply executor/split.go]: done.

File: executor/split.go (hunk 12 above)
Comment (crazycs520) [reply executor/split.go]: Great, thanks.

**Pull request merged**